File: scripts/wasm-source-fingerprint.py
#!/usr/bin/env python3
"""Fingerprint committed engine inputs without frontend consumer directories."""

import hashlib
import subprocess


CONSUMERS = {
    b"rau-credits",
    b"rhwp-agent",
    b"rhwp-chrome",
    b"rhwp-firefox",
    b"rhwp-safari",
    b"rhwp-shared",
    b"rhwp-studio",
    b"rhwp-vscode",
}

tree = subprocess.check_output(["git", "ls-tree", "-z", "HEAD:rhwp"])
entries = (
    entry
    for entry in tree.split(b"\0")
    if entry and entry.split(b"\t", 1)[1] not in CONSUMERS
)
digest = hashlib.sha256()
for entry in entries:
    digest.update(entry + b"\0")
print(digest.hexdigest())

File: scripts/wasm-source-fingerprint.mjs
import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';

// These apps consume the engine; none is a Rust build input. Everything else,
// including new engine directories and embedded assets, invalidates the cache.
const consumers = new Set([
  'rau-credits', 'rhwp-agent', 'rhwp-chrome', 'rhwp-firefox',
  'rhwp-safari', 'rhwp-shared', 'rhwp-studio', 'rhwp-vscode',
]);
const tree = execFileSync('git', ['ls-tree', '-z', 'HEAD:rhwp'], { encoding: 'utf8' });
const digest = createHash('sha256');
for (const entry of tree.split('\0')) {
  if (entry && !consumers.has(entry.slice(entry.indexOf('\t') + 1))) {
    digest.update(entry + '\0');
  }
}
console.log(digest.digest('hex'));
